# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/spatial_tree.rs
# Commit: b04340bad99d
# Full Hash: b04340bad99d70638efda2e7d528e89df77e47c4
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2020-05-22 09:43:16
# Description:
#   Bug 1617524 - Fix crash in get_relative_transform edge case. r=jnicol
#   
#   Previously, WR would attempt to detect at the start of frame
#   building if the spatial node of any picture cache contained
#   a non-axis-aligned transform, and disable picture caching in
# ==============================================================================

diff -r fdfb16cb82a6 -r b04340bad99d gfx/wr/webrender/src/spatial_tree.rs
--- a/gfx/wr/webrender/src/spatial_tree.rs	Thu May 21 18:23:11 2020 +0000
+++ b/gfx/wr/webrender/src/spatial_tree.rs	Thu May 21 07:42:32 2020 +0000
@@ -631,15 +631,30 @@
         while node_index != ROOT_SPATIAL_NODE_INDEX {
             let node = &self.spatial_nodes[node_index.0 as usize];
             match node.node_type {
-                SpatialNodeType::ReferenceFrame(..) |
-                SpatialNodeType::StickyFrame(..) => {
-                    // TODO(gw): In future, we may need to consider sticky frames.
+                SpatialNodeType::ReferenceFrame(ref info) => {
+                    match info.kind {
+                        ReferenceFrameKind::Zoom => {
+                            // We can handle scroll nodes that pass through a zoom node
+                        }
+                        ReferenceFrameKind::Transform |
+                        ReferenceFrameKind::Perspective { .. } => {
+                            // When a reference frame is encountered, forget any scroll roots
+                            // we have encountered, as they may end up with a non-axis-aligned transform.
+                            scroll_root = ROOT_SPATIAL_NODE_INDEX;
+                        }
+                    }
                 }
+                SpatialNodeType::StickyFrame(..) => {}
                 SpatialNodeType::ScrollFrame(ref info) => {
-                    // If we found an explicit scroll root, store that
-                    // and keep looking up the tree.
-                    if let ScrollFrameKind::Explicit = info.frame_kind {
-                        scroll_root = node_index;
+                    match info.frame_kind {
+                        ScrollFrameKind::PipelineRoot => {
+                            // Once we encounter a pipeline root, there is no need to look further
+                            break;
+                        }
+                        ScrollFrameKind::Explicit => {
+                            // Store the explicit scroll root, keep looking up the tree
+                            scroll_root = node_index;
+                        }
                     }
                 }
             }